### PR TITLE
Rewrite github project references

### PR DESCRIPTION
**Pull request by pelikhan**
Fix for https://github.com/microsoft/pxt/issues/6136

This change handles gracefully the publishing of github project with local references. 

When a project gets published, 
* all github local dependencies (workspace:*) get rewritten to github references (github:*)
* the tag is used otherwise the branch is used
* if the commit is used to create tag, tag requirement is enforced.

Scenario:

repoA and repoB are both used locally. repoB has a dependency on repoA. When commiting, reference to repoA workspace:... should be converted to github:...

### CODE REVIEW

File: webapp/src/data.ts [hunk 1]
```diff
@@ -122,7 +122,7 @@ function unsubscribe(component: AnyComponent) {
 
 function expired(ce: CacheEntry) {
     if (!ce.api.expirationTime)
-        return ce.data != null;
+        return !ce.lastRefresh; // needs to be refreshed at least once
```
Comment (pelikhan) [inline webapp/src/data.ts]: stumbled on this one: if a api is Async with no expirationTime, it is never called.

File: webapp/src/package.ts [hunk 2]
```diff
@@ -665,9 +674,30 @@ data.mountVirtualApi("open-pkg-meta", {
...
             files.filter(f => !!f.diagnostics).forEach(f => r.diagnostics.concat(f.diagnostics));
         }
-        const hasGit = f.getPkgId() == "this" && !!(f.header && f.header.githubId)
-        if (hasGit)
-            r.numFilesGitModified = files.filter(f => f.baseGitContent != f.content).length;
+        return r;
+    }
+})
+
+export interface PackagetGitStatus {
```
Comment (jwunderl) [inline webapp/src/package.ts]:
```suggestion
export interface PackageGitStatus {
```

Think I got all the places it's refd

File: webapp/src/package.ts [hunk 3]
```diff
@@ -665,9 +674,30 @@ data.mountVirtualApi("open-pkg-meta", {
...
+}
+
+/*
+    pkg-git-status:<guid>
+*/
+data.mountVirtualApi("pkg-git-status", {
+    getSync: p => {
+        p = data.stripProtocol(p)
+        const f = allEditorPkgs().find(pkg => pkg.header && pkg.header.id == p);
+        const r: PackagetGitStatus = {};
```
Comment (jwunderl) [inline webapp/src/package.ts]:
```suggestion
        const r: PackageGitStatus = {};
```

File: webapp/src/githubbutton.tsx [hunk 4]
```diff
@@ -59,10 +59,8 @@ export class GithubButton extends sui.UIElement<GithubButtonProps, GithubButtonS
                 onClick={this.createRepository} />
 
         // existing repo
-        //const targetTheme = pxt.appTarget.appTheme;
-        const mainPkg = pkg.mainEditorPkg()
-        const meta: pkg.PackageMeta = this.getData("open-pkg-meta:" + mainPkg.getPkgId());
-        const modified = meta && !!meta.numFilesGitModified;
+        const meta: pkg.PackagetGitStatus = this.getData("pkg-git-status:" + header.id);
```
Comment (jwunderl) [inline webapp/src/githubbutton.tsx]:
```suggestion
        const meta: pkg.PackageGitStatus = this.getData("pkg-git-status:" + header.id);
```

**Comment by pelikhan** [comment]
It does restore the workspace reference so you still reference the local projects in your browser but have GitHub refs in GitHub.

**Pull request merged**